cpp/src/arrow/compute/kernels/vector_swizzle.cc

```diff
@@ -0,0 +1,422 @@
+  }
+
+  template <typename Type>
+  enable_if_t<is_integer_type<Type>::value, Status> Visit(const Type& output_type) {
+    using OutputCType = typename Type::c_type;
+
+    RETURN_NOT_OK(CheckInput(output_type));
+
+    // Dispatch the execution based on whether there are likely many nulls in the output.
+    // - If many nulls (i.e. the output is "sparse"), preallocate an all-false validity
+    // buffer and a zero-initialized data buffer (just to avoid exposing previous memory
+    // contents - even if it is shadowed by the validity bit). The subsequent processing
+    // will fill the valid values only.
+    // - Otherwise (i.e. the output is "dense"), the validity buffer is lazily allocated
+    // and initialized all-true in the subsequent processing only when needed. The data
+    // buffer is preallocated and filled with "impossible" values (that is, input_length -
+    // note that the range of inverse_permutation is [0, input_length)) for the subsequent
+    // processing to detect validity.
+    if (LikelyManyNulls()) {
+      RETURN_NOT_OK(AllocateValidityBufAndFill(false));
+      RETURN_NOT_OK(AllocateDataBufAndZero(output_type));
+      return Execute<Type, true>();
+    } else {
+      RETURN_NOT_OK(
+          AllocateDataBufAndFill(output_type, static_cast<OutputCType>(input_length_)));
+      return Execute<Type, false>();
+    }
+  }
+
+  Status Visit(const DataType& output_type) {
+    DCHECK(false) << "Shouldn't reach here";
+    return Status::Invalid("Shouldn't reach here");
+  }
+
+ private:
+  KernelContext* ctx_;
+  const ShapeType& indices_;
+  const int64_t input_length_;
+  const int64_t output_length_;
+
+  std::shared_ptr<Buffer> validity_buf_;
+  std::shared_ptr<Buffer> data_buf_;
+
+  InversePermutationImpl(KernelContext* ctx, const ShapeType& indices,
+                         int64_t input_length, int64_t output_length)
+      : ctx_(ctx),
+        indices_(indices),
+        input_length_(input_length),
+        output_length_(output_length) {}
+
+  template <typename Type>
+  Status CheckInput(const Type& output_type) {
+    using OutputCType = typename Type::c_type;
+
+    if (static_cast<int64_t>(std::numeric_limits<OutputCType>::max()) < input_length_) {
+      return Status::Invalid(
+          "Output type " + output_type.ToString() +
+          " of inverse_permutation is insufficient to store indices of length " +
+          std::to_string(input_length_));
+    }
+
+    return Status::OK();
+  }
+
+  bool LikelyManyNulls() { return output_length_ > 2 * input_length_; }
+
+  Status AllocateValidityBufAndFill(bool valid) {
+    DCHECK_EQ(validity_buf_, nullptr);
+
+    ARROW_ASSIGN_OR_RAISE(validity_buf_,
+                          AllocateEmptyBitmap(output_length_, ctx_->memory_pool()));
+    auto validity = validity_buf_->mutable_data_as<uint8_t>();
+    std::memset(validity, valid ? 0xff : 0, validity_buf_->size());
```

Comment:
Note: if allocating an uninitialized buffer, should use `validity_buf_->capacity()` to make sure that padding bits are initialized too.

Reply:
Very useful advice! Addressed.